iml-gui/crate/src/components/command_modal.rs

```diff
@@ -284,3 +537,601 @@ fn close_button() -> Node<Msg> {
+                .iter()
+                .filter(|x| is.contains(&x.id))
+                .map(|x| x.clone())
+                .collect()
+        }
+        fn select_jobs(&self, is: &[u32]) -> Vec<Job0> {
+            self.all_jobs
+                .iter()
+                .filter(|x| is.contains(&x.id))
+                .map(|x| x.clone())
+                .collect()
+        }
+        fn select_steps(&self, is: &[u32]) -> Vec<Step> {
+            self.all_steps
+                .iter()
+                .filter(|x| is.contains(&x.id))
+                .map(|x| x.clone())
+                .collect()
+        }
+    }
+
+    #[test]
+    fn test_parse_job() {
+        assert_eq!(extract_uri_id::<Job0>("/api/job/39/"), Some(39));
+        assert_eq!(extract_uri_id::<Step>("/api/step/123/"), Some(123));
+        assert_eq!(extract_uri_id::<Command>("/api/command/12/"), Some(12));
+        assert_eq!(extract_uri_id::<Command>("/api/xxx/1/"), None);
+    }
+
+    #[test]
+    fn test_is_subset() {
+        let all = vec![1, 2, 3, 4, 5];
+        assert_eq!(is_subset(&vec![1, 2, 3], &all), true);
+        assert_eq!(is_subset(&vec![1, 3, 5], &all), true);
+        assert_eq!(is_subset(&vec![], &all), true);
+        assert_eq!(is_subset(&all, &all), true);
+
+        assert_eq!(is_subset(&vec![1, 6], &all), false);
+        // if not sorted, the correctness is not guaranteed
+        assert_eq!(is_subset(&vec![5, 1], &all), false);
+    }
+
+    #[test]
+    fn test_interpret_click() {
+        // test transitions, we simulate the clicks on various elements
+
+        let pair = (Select::Command(77), false);
+
+        let pair = interpret_click(&pair.0, TypedId::Job(162));
+        assert_eq!(&pair, &(Select::CommandJob(77, vec![162]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Job(163));
+        assert_eq!(&pair, &(Select::CommandJob(77, vec![162, 163]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Step(309));
+        assert_eq!(&pair, &(Select::CommandJobSteps(77, vec![162, 163], vec![309]), true));
+
+        let pair = interpret_click(&pair.0, TypedId::Step(308));
+        assert_eq!(
+            &pair,
+            &(Select::CommandJobSteps(77, vec![162, 163], vec![308, 309]), true)
+        );
+
+        let pair = interpret_click(&pair.0, TypedId::Job(163));
+        assert_eq!(&pair, &(Select::CommandJobSteps(77, vec![162], vec![308, 309]), false));
+
+        let pair = interpret_click(&pair.0, TypedId::Job(162));
+        assert_eq!(&pair, &(Select::Command(77), false));
+    }
+
+    /// There could the the problem, when `schedule_fetch_tree` made the api requests
+    /// in one order, but during delays or something the responses may come in different
+    /// order. In all such cases, however, model should remain consistent,
+    /// and the test checks this.
+    #[test]
+    fn test_async_handlers_consistency() {
+        let mut rng = Xoroshiro64Star::seed_from_u64(485369);
```

Comment:
These are not really random (even pseudo) numbers, are they?

Reply:
It is pseudo random number generator with pretty long period (2^64), sufficient for the testing purposes.

Reply:
I was trying to tell that you are using a fixed seed and are getting the same sequence of numbers each time. There is no more randomness than in https://xkcd.com/221/ :)

Reply:
I promise, I fairly diced the roll 6 times!

Well, I am sure the tests should be deterministic. If I seeded it from entropy (if I somehow overlooked a bug), then it would fail sometimes in the CI, which might mess some other's work. And this work might be more important than the bug, that we might never normally see.